### PR TITLE
Maintain changedVariablesStack from DPath VRef

### PR DESCRIPTION
**Pull request by mbeckerle**
DAFFODIL-2565

### CODE REVIEW

**Comment by mbeckerle** [comment]
I actually believe the "fix" in this PR is wrong. The code was correct before. Just a little confusing.
PState contains the changedVariablesStack, which must be maintained whenever variables are set or read or newInstances created.

The variableMap.readVariable method was doing a callback to PState to maintain this changedVariablesStack. That means all uses of readVariable, including from DPath VRef, were doing this maintenance.

However, variableMap.setVariable does no such callback.

Instead, code that is setting variables is supposed to call the PState.setVariable method. which does this changedVariablesStack maintenance.

The reality: There's more state to maintain having to do with variables than just the variableMap, so all manipulation of variables SHOULD go by way of PState/UState methods, and those should delegate things to the VariableMap object, and maintain other things like changedVariablesStack. 

So the change so that nobody uses variableMap directly, and instead uses variables by way of PState/UState is the way to go, and the uses of variableMap directly need to go away, and the callback from variableMap to PState also should go away.

**Review by mbeckerle — COMMENTED** [review]
fix needed.

**Comment by mbeckerle** [comment]
So at this point, the only thing which directly uses the VariableMap object is the debugger, and that's because it wants to display the state transitions of the variables. So variableMap can't be a private member of ParseOrUnparseState, but it's better isolated.

**Review by stevedlawrence — APPROVED** [review]
+1, but are we missing tests?

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala [hunk 1]
```diff
@@ -74,6 +76,21 @@ abstract class UState(
   extends ParseOrUnparseState(vbox, diagnosticsArg, dataProcArg, tunable)
   with Cursor[InfosetAccessor] with ThrowsSDE with SavesErrorsAndWarnings {
 
+  final override def setVariable(vrd: VariableRuntimeData, newValue: DataValuePrimitive, referringContext: ThrowsSDE) =
+    vbox.vmap.setVariable(vrd, newValue, referringContext, this)
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala:80]: Missed a spot. 
SetVariableSuspendableExpression is directly using the variableMap. 

I am going to make the variableMap private in PState/UState, but let the debugger grab it via a variableMapForDebugger method but everything else will have to access it only via the PState/UState methods.

[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala:82]: Latest commit should fix the test coverage issue. It enforces that all get/set of variables is done via PState/UState uses only.

**Review by jadams-tresys — APPROVED** [review]
+1

This is a nice clean up of a lot of the stuff that I had kludged together to get newVariableInstance working.

**Pull request merged**